operators/operating-pinot/consistent-push.md

```diff
@@ -0,0 +1,94 @@
+
+### How this works
+
+Segment lineage data structure has been introduced in Zookeeper (under the path `<cluster_name>/PROPERTYSTORE/SEGMENT_LINEAGE/<table_name>`) for keeping track of which segments have been replaced by which new set of segments, as well as corresponding state and timestamp.
+
+```
+{
+  "id": "<table_name>",
+  "simpleFields": {},
+  "mapFields": {},
+  "listFields": {
+    "<segment_lineage_entry_id>": [
+      "<segmentsFrom_list>",
+      "<segmentsTo_list>",
+      "<state>",
+      "<timestamp>"
+    ]
+  }
+}
+```
+
+When broker answers queries from the users, it will go through the lineage entries and only route to the segments in `segmentsFrom` for those in "IN_PROGRESS" or "REVERTED" state and 
+the segments in `segmentsTo` for those in "COMPLETED" state, therefore preserving data snapshot atomicity.
+
+Below are the APIs available on the controller to invoke the segment replacement protocol.
+
+1. `startReplaceSegments`: Signifies to the controller that a replacement protocol is about to atomically replace `segmentsFrom`, a source list of segments, by `segmentsTo` , a target list of segments, which then persists a segment lineage entry with "IN PROGRESS" state to Zookeeper and returns its ID. 
+2. `endReplaceSegments`: Ends the replacement protocol associated with the segment lineage entry ID passed in as a parameter by changing the state to "COMPLETED".
+3. `revertReplaceSegments`: Reverts the replacement protocol associated with the segment lineage entry ID passed in as a parameter by changing the state to "REVERTED".
+
+However, we don't typically expect users to invoke these APIs directly.
+
+Instead, consistent push is built into batch ingestion jobs (currently only supported for the standalone execution framework).
+
+### How to set up Ingestion Job with Consistent Push
+
+**Step 0**: Set up config for your OFFLINE, REFRESH table. Enable `consistentDataPush` under IngestionConfig -> BatchIngestionConfig.
+
+```
+"tableName": "myTable_OFFLINE",
+"tableType": "OFFLINE",
+...
+...
+"ingestionConfig": {
+  "batchIngestionConfig": {
+    "segmentIngestionType": "REFRESH",
+     "segmentIngestionFrequency": "DAILY", // or HOURLY
+     "consistentDataPush": true
+  }
+}
+```
+
+**Step 1**: Execute the job as usual by following https://docs.pinot.apache.org/users/tutorials/batch-data-ingestion-in-practice#executing-the-job
+
+### How to trigger Data Rollback
+
+**Step 0**: Identify the segment lineage entry ID corresponding to the segment swap that would like to be rolled back. This could be identified by heading to the Zookeeper browser and inspecting under the path: `<cluster_name>/PROPERTYSTORE/SEGMENT_LINEAGE/<table_name>`.
+
+**Step 1**: Head to Swagger REST API to utilize the `revertReplaceSegments` API to rollback data.
+
+### Cleanup
```

Comment:
It would be good to add the cleanup in startReplaceSegment api also:
1. We proactively remove the first snapshot if the client side is pushing the 3rd snapshot, so we are not exceeding the 2x disk space.
2. If the previous push fails in the middle (IN_PROGRESS/REVERTED state), we also clean up the segmentsTo.